Correction of Crash in Indoor Pool Model Related to Improper Surface Reference

RKStrand opened the pull request:
Pull request overview
---------------------
 - Defect: Fixes #7986 
 - This pull request fixes a problem in the indoor swimming pool model.  The user file that highlighted this problem also had a user error in it.  The user defined a swimming pool, but the surface that was associated with this pool was not a valid floor (in fact, it wasn't any surface in the IDF).  This generates an error.  However, the code logic in EnergyPlus allowed the algorithm to get to a line that ended up trying to access a variable with an index of zero.  This caused a hard crash as a result.  The fix avoids this hard crash by moving the offending line of code to its proper position and moved the various checks on the surface to a new subroutine.  A unit test to exercise this new subroutine was added.  In addition, some text was added to the IO Ref section for swimming pools that specifically outline what the rules are for a valid surface that is the pool.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Correction of hard crash in pool model

This commit fixes the hard crash that takes place when the user chooses a pool surface that does not exist in the input file.  This results in a surface pointer of zero which then causes a bad indexing problem because the index is zero.  This is now fixed and a new separate subroutine was created for the various checks made on the pool surface.

RKStrand committed:
Unit Test and Documentation

A unit test for the bug fix was created based on the new subroutine that was developed as part of this work.  In addition, clarifying documentation was added to the Input Output Reference so that it is clear what type of surface is and is not allowed for the modeling of pools in EnergyPlus.

RKStrand committed:
Merge branch 'develop' into 7986SwimmingPoolIndoorCrashWithIncorrectSurface

rraustad commented on tst/EnergyPlus/unit/SwimmingPool.unit.cc:
```diff
@@ -255,3 +257,127 @@ TEST_F(EnergyPlusFixture, SwimmingPool_InitSwimmingPoolPlantNodeFlow)
+    // Tests for InitSwimmingPoolPlantLoopIndex
+    SwimmingPool::clear_state();
+    DataSurfaces::clear_state();
+    DataHeatBalance::clear_state();
```
Clear_states do not belong in unit tests. They are handled by the test fixture.

RKStrand replied:
@rraustad Got it.  Removed and I'll push that commit up once energyplus_test runs successfully.

RKStrand replied:
Done.  These lines have been removed.  Energyplus_test still runs completely as expected with no errors.  A new commit has now been made for these changes.

rraustad commented on src/EnergyPlus/SwimmingPool.cc:
```diff
@@ -460,6 +420,61 @@ namespace SwimmingPool {
+                ErrorsFound = true;
+            }
+        }
+    }
```
Nice, this is a very complete unit test. Where do you check if the surface is referenced by another pool? I didn't look but is that done after all pools are read in?

RKStrand replied:
@rraustad It's "Test 2" where PartOfVentSlabOrRadiantSurface is set to true.  That also tracks pool as well as ventilated slabs and radiant systems.  That variable is set to true in the new subroutine when all the checks pass in the "else" of the if block.

rraustad replied:
So a pool is a ventilated slab? or is it a radiant surface? Yes, I do see in the new function where this var is set true for pools.

    } else { // ( Pool( Item ).SurfacePtr > 0 )
       DataSurfaces::Surface(Pool(Item).SurfacePtr).PartOfVentSlabOrRadiantSurface = true;

I see 14 matches. What do you think of PartOf(or Is)VentSlabOrRadSurfOrPool ?? Won't hold you to this, but I was confused, so will another developer in the future also be confused?

I'm not sure this answers my question. If I have an input file with 2 pools and I specify the same surface name for both, is there an error?

Myoldmopar replied:
I agree, this is a great example of extracting out a reasonable amount of code into a standalone function that is easily tested.  One step that could've taken this over the top would've been to make this function a method on the pool struct so that inside the function you never had to look up `Pool(Item)` like you do a bunch here.  If you do end up needing to do anything else to resolve @rraustad 's comments here, I would suggest you take a couple minutes and make that change.  If no more commits are needed, don't sweat it.

RKStrand replied:
@rraustad @Myoldmopar Ok, I've committed changes to address this request by @Myoldmopar and brought the code up to speed with develop.  I think this is now good to go though I understand that it might need another round of reviews given the number of changes that I made.  The changes should be algorithm neutral (except for the elimination of an unused variable).  Let me know if there's anything else that needs to be done here.  I am hoping that CI will come back clean and this is ready to go.

RKStrand committed:
Change to address review comment

PR reviewer made a comment that required a change in the unit test.  This addresses that comment.

RKStrand committed:
Change of variable name

One of the review comments requested that a name change take place for a variable related to the original bug fix.  Since the previous name only referenced radiant surfaces or ventilated slabs and not also pools, there could have been some confusion for future developers.  The name has been updated to reflect all three.  Basically, any particular surface can only be a radiant system, a ventilated slab, or a pool, but not multiples of those things at the same time.

RKStrand committed:
Merge branch 'develop' into 7986SwimmingPoolIndoorCrashWithIncorrectSurface

rraustad commented:
@Myoldmopar I think Mac CI is a little grumpy.

rraustad commented:
Oh, wait, that was the previous run (13 hours ago). This one looks OK.

RKStrand commented:
@rraustad Yeah, there have occasionally been some blips.  The latest commits earlier today include the name change that you requested and a merge of develop into this branch.

@Myoldmopar I'm going to start working on the other recommendations that you had.  Hoping that goes smoothly but at least if things go horribly wrong we can always back things up (beep...beep...beep) and still be okay.

Myoldmopar commented:
@rraustad yep, Mac is finally looking good.  It is showing a warning, but it's actually a valid warning I'm going to fix in a small PR shortly.  Otherwise I think it's finally settled down.  And I have access to the machine again, so I can kick it if it needs it sometime.

RKStrand committed:
Improvement of coding

Additional review request for some code modernization has been completed.  An unused variable was also discovered and eliminated in the process.

RKStrand committed:
Merge branch 'develop' into 7986SwimmingPoolIndoorCrashWithIncorrectSurface

Myoldmopar commented on src/EnergyPlus/SwimmingPool.cc:
```diff
@@ -226,10 +226,6 @@ namespace SwimmingPool {
-        for (Item = 1; Item <= NumSwimmingPools; ++Item) {
-            Pool(Item).SurfaceToPoolIndex.allocate(DataSurfaces::TotSurfaces);
-            Pool(Item).SurfaceToPoolIndex = 0;
-        }
```
A thousand times yes.

Myoldmopar commented on src/EnergyPlus/SwimmingPool.cc:
```diff
@@ -460,6 +416,59 @@ namespace SwimmingPool {
+            ShowSevereError(RoutineName + "Invalid " + cAlphaField2 + " = " + Alpha2);
+            ShowContinueError("Occurs in " + CurrentModuleObject + " = " + Alpha1);
+            ErrorsFound = true;
+        } else if (DataSurfaces::Surface(this->SurfacePtr).IsRadSurfOrVentSlabOrPool) {
```
This is really a great addition.  

While the reduced lookups have the potential for a miniscule speedup in this case (it can when it is applied more widespread: https://github.com/NREL/EnergyPlus/pull/4192#issuecomment-39848360), more importantly it reduces the amount of code for a developer to process.  

I would suggest in the future taking it one step farther and creating a reference to the surface of interest.  You'll want to first check if the surface <= 0 and leave early if it is:
```
if (this->SurfacePtr <= 0) {
    ShowSevereError(RoutineName + "Invalid " + cAlphaField2 + " = " + Alpha2);
    ShowContinueError("Occurs in " + CurrentModuleObject + " = " + Alpha1);
    ErrorsFound = true;
    return;
}
```

Then after this point, you can comfortably get a reference to a surface:

```
auto & surface = DataSurfaces::Surface(this->SurfacePtr);
```

You can then replace all the instances of `DataSurfaces::Surface(this->SurfacePtr)` with `surface`.  Once again, this isn't really a performance thing, but more a cognitive complexity overhead.  When a developer reads this code, there are still a bunch of array lookups going on amidst all the actual logic and functionality.  Cleaning up the code in ways like this can make the function's intention much more clear.

Myoldmopar commented on src/EnergyPlus/SwimmingPool.hh:
```diff
@@ -162,6 +161,12 @@ namespace SwimmingPool {
+                                        std::string const Alpha2,
+                                        std::string const cAlphaField2,
+                                        bool &ErrorsFound
+        );
```
Another minor nitpick for later.  Try to start naming methods with `camelCase` instead of `PascalCase`.  Class names and namespaces should be `PascalCase`, but all variables and functions should be `camelCase`.

Myoldmopar commented on tst/EnergyPlus/unit/SwimmingPool.unit.cc:
```diff
@@ -255,3 +257,120 @@ TEST_F(EnergyPlusFixture, SwimmingPool_InitSwimmingPoolPlantNodeFlow)
+    EXPECT_TRUE(DataSurfaces::Surface(poolReference.SurfacePtr).IsPool);
+    EXPECT_EQ(DataSurfaces::Surface(poolReference.SurfacePtr).Zone,poolReference.ZonePtr);
+    
+}
```
Great unit test.  You could also use the `auto & surface` reference approach here to reduce the apparent complexity of this code, but it's not super necessary here in unit test land.

Myoldmopar reviewed (approved):
I am happy with this.  I could nitpick all day, but this is a great defect fix that also introduces some code cleanups including object orientation.  Great work @RKStrand, thanks for adding that to this defect fix.

Myoldmopar commented:
Windows and Linux are wrapping up on this PR and then it will drop in.

Myoldmopar commented:
Hmmm, Windows didn't like something.  But there aren't any reasons given.  Why tf can't CI just be cool man? :sunglasses:

Myoldmopar commented:
Hmm, suspicious that Windows would fail to run unit tests around the same time as @lgu1234 started reporting a compiler error in the unit test code.  OK, I don't think I'll hold up this PR for another round, but let me at least verify before merging this.

Myoldmopar commented:
I pulled this branch on my Windows VM.  I was able to again confirm the build error in the unit test executable.  I'm not quite sure why the error message wasn't caught by CI but I believe that is really the underlying problem with the test failures here.  I then pulled develop into this branch, as develop now has the unit test fix.  I was then able to directly build the unit test binary successfully.  I'm going to merge this in without waiting on CI.
